contracts/RewardsPoolBase.sol

```diff
@@ -271,6 +373,35 @@ contract RewardsPoolBase is ReentrancyGuard {
         return user.rewardDebt.length;
     }
 
-    // TODO add logic for reward per block manipulation
-    // TODO add logic for extension
+    /**
+        @dev Extends the rewards period and updates the rates
+        @param _endBlock  new end block for the rewards
+        @param _rewardsPerBlock array with new rewards per block for each token 
+     */
+    function extend(uint256 _endBlock, uint256[] memory _rewardsPerBlock)
+        public
+    {
+        require(_endBlock > _getBlock(), "End block must be in the future");
+        require(
+            _endBlock >= endBlock,
+            "End block must be after the current end block"
+        );
+        require(
+            _rewardsPerBlock.length == rewardsTokens.length,
+            "Rewards amounts length is less than expected"
+        );
+
+        for (uint256 i = 0; i < _rewardsPerBlock.length; i++) {
+            rewardPerBlock[i] = _rewardsPerBlock[i];
+
+            IERC20Detailed(rewardsTokens[i]).safeTransferFrom(
+                msg.sender,
+                address(this),
+                _rewardsPerBlock[i]
+            );
+        }
+
+        endBlock = _endBlock;
+        updateRewardMultipliers();
```

Comment:
This should happen before the updates of the `rewardPerBlock`. Otherwise you are also taking away the reward accrued based on the old reward since the last update.